app/services/friend_requests/accept.rb

```diff
@@ -0,0 +1,21 @@
+module FriendRequests
+  class Accept
+    include Serviceable
+
+    def initialize(friendship:)
+      @friendship = friendship
+    end
+
+    def call
+      @friendship.update!(
+        mutual: true,
+      )
+
+      Friendship.create!(
+        user:   @friendship.friend,
+        friend: @friendship.user,
+        mutual: true,
+      )
+    end
```

Comment:
This is where `mutual` gets set to true when the inverse friendship is created.

Reply:
💡 Growing my OOP brain. In comparison, in Elixir, I would do this as a `with` statement in a context function, which I can now see as kind of analogous to this approach

But just to make sure that I have this right: the main advantage here is that this operation doesn't really belong in the model or controller, so this module encapsulates this functionality. As of now it primarily sets `mutual` to true, but would make more complex operations simpler to handle in the future. Did I get that right?

Reply:
Yes, to elaborate, here are our options, as I see it: 
1. **Bespoke-Endpoint Controller Logic**
   Keep record create, update, etc. business logic in controller. For example, take code from L10-18 and just pop it in the controller instead.
1. **Rails Callbacks**
   Use `after_create` or similar callback in order to make these modifications when certain conditions are fulfilled. For example, we could do something like when a `Friendship` is created, if the inverse `Friendship` is `mutual: true`, then make this Friendship `mutual: true` as well. 
1. **Model-Centralized**
   L10-18 above lives in some method "create_inverse" on the Friendship.
1. **Service Abstraction**
   This is the above solution, where some intermediate Service is responsible for the above code. In this solution, it's assumed that we will ALWAYS use this service to do the described function, in this case, accepting a friend request. 

Here are the pros/cons of each pattern, as I see it:

1. **Bespoke-Endpoint Controller Logic**
   - **Benefits**: 
     - Clear from a bird's-eye view at the controller what the endpoint actually does - no surprises.
   - **Problems**:
     - Reusing code in between controllers becomes impossible, since calling 1 endpoint from another is a bad practice. Very duplicative. If the logic has any value as a reusable function, this is not the best solution. 
     - Heavy or dense endpoints tend to have poor performance. 
     - Heavy or dense endpoints are hard to read.
1. **Rails Callbacks**
   - **Benefits**: 
     - Reusable.
     - Honestly, pretty elegant! Usually quite dry and concise. 
     - Don't need to worry about always using a single pathway, no matter what pathway or endpoint you use, the necessary operations will take place.
     - The only "event driven" equivalent in Rails.
   - **Problems**:
     - AGGRESSIVELY reusable, i.e. it will implicitly apply to every create callback thereafter. 
     - Separates business logic out from other centralized areas. Can't tell you how many times I've been debugging a rails app and digging through the entire controller flow only to find out that whatever thing has been automatically and silently happening due to a callback firing. Breeds inconsistency in terms of approach.
       - https://www.reddit.com/r/rails/comments/15thm2z/comment/jwkqtgl explains this take pretty well. 
     - Relies on implicit behavior firing "around" our endpoint behavior instead of being a fundamental part of our flow.
     - Unclear from a bird's eye perspective what is actually happening at each stage. Becomes messy easily. 
1. **Model-Centralized**
   - **Benefits**: 
     - Reusable.
     - Always know where to find business logic for a given model.
   - **Problems**:
     - Will end up with many methods that are only called in particular flows, can be hard to determine when a particular thing is called or why. 
     - By "belonging" to the model, it is awkward to adjust any other records outside of the model or do any cross-table functionality, and gets sketchy when we need to cascade changes to associated models from the parent model itself. This leads to lots of invisible behavior. Consider an endpoint that gives a number of records a rank. Putting the `rerank` code on the record itself means that behavior on a single record ends up touching many others.  
1. **Service Abstraction**
   - **Benefits**: 
     - Reusable, but explicit, functionality. We are never accidentally triggering behavior.
     - Encourages single entry/exit point to some of these behaviors. For example, if there are other areas in our app where we want to accept a friend request, it will be easy to reuse. 
     - Less weird to touch multiple records or models, since we aren't saying "this functionality belongs to a single Friendship object", and instead we essentially develop a Friendship Request Manager. 
   - **Problems**:
     - End up with a thousand service objects that are MOSTLY wrapping the underlying models with behavior.
     - A little more onerous and less dry, we end up repeating the Service initializaiton pattern many times.

Reply:
Awesome! Thanks!

Great writeup, really nailed it down for me. I'm fully on board with the services architecture, I think it gives us a very clear design path to follow moving forward.